lib/optimizely/custom_attribute_condition_evaluator.rb

```diff
@@ -0,0 +1,166 @@
+      #                    2) the user attribute value is neither null nor undefined
+      #                 Returns false otherwise
+
+      return false unless @user_attributes
+      !@user_attributes[condition['name']].nil?
+    end
+
+    def greater_than_evaluator(condition)
+      # Evaluate the given greater than match condition for the given user attributes.
+      # Returns boolean true if the user attribute value is greater than the condition value,
+      #                 false if the user attribute value is less than or equal to the condition value,
+      #                 nil if the condition value isn't a number or the user attribute value isn't a number.
+
+      condition_value = condition['value']
+      user_provided_value = @user_attributes[condition['name']]
+
+      return nil unless user_provided_value.is_a?(Numeric) && condition_value.is_a?(Numeric)
+
+      return nil if infinite_number?(user_provided_value) || infinite_number?(condition_value)
+
+      user_provided_value > condition_value
+    end
+
+    def less_than_evaluator(condition)
+      # Evaluate the given less than match condition for the given user attributes.
+      # Returns boolean true if the user attribute value is less than the condition value,
+      #                 false if the user attribute value is greater than or equal to the condition value,
+      #                 nil if the condition value isn't a number or the user attribute value isn't a number.
+
+      condition_value = condition['value']
+      user_provided_value = @user_attributes[condition['name']]
+
+      return nil unless user_provided_value.is_a?(Numeric) && condition_value.is_a?(Numeric)
+
+      return nil if infinite_number?(user_provided_value) || infinite_number?(condition_value)
+
+      user_provided_value < condition_value
+    end
+
+    def substring_evaluator(condition)
+      # Evaluate the given substring match condition for the given user attributes.
+      # Returns boolean true if the condition value is a substring of the user attribute value,
+      #                 false if the condition value is not a substring of the user attribute value,
+      #                 nil if the condition value isn't a string or the user attribute value isn't a string.
+
+      condition_value = condition['value']
+      user_provided_value = @user_attributes[condition['name']]
+
+      return nil unless user_provided_value.is_a?(String) && condition_value.is_a?(String)
+
+      user_provided_value.include? condition_value
+    end
+
+    private
+
+    def value_valid_for_exact_conditions?(value)
+      # Returns true if the value is valid for exact conditions. Valid values include
+      #  strings, booleans, and numbers that aren't NaN, -Infinity, or Infinity.
+
+      return value.finite? if value.is_a? Float
+
+      (value.is_a? FalseClass) || (value.is_a? Integer) || (value.is_a? String) ||
```

Comment:
don't we need a finite check with integer class when we need it for float?

Reply:
Ruby throws ZeroDivisionError against integer division.
Class infinity works with float.